### PR TITLE
Prp refactoring

### PR DESCRIPTION
**Pull request by robertavram**
Base branch for PRP Refactoring work.

### Step 1:
- [x] +Endpoint for listing/adding new Intervention ResultLink `interventions/<intervention_pk>/result-links/`
- [x] (Philip) Test endpoint for adding new Result Links (#923)
- [x] + Endpoint for getting/patching/deleting ResultLinks `interventions/result-links/<pk>/`
- [x] (Philip) Test endpoint for patching and deleting ResultLinks (#947)
- [x] + Endpoint for listing/adding new LowerResult (PD Output) `interventions/result-links/<result_link_pk>/lower-results/`
- [x] (Philip) Test endpoint for listing/adding new LowerResult (#951)
- [x] + Endpoint for listing/adding new AppliedIndicator 'interventions/lower-results/<lower_result_pk>/indicators/`
- [x] (Philip) Test Endpoint for listing/adding new AppliedIndicators (#960)

- [x] (Cory) Endpoint for getting/patching/deleting disaggregations (disabled deleting) (https://github.com/unicef/etools/pull/919)
- [x] (Cory) Test endpoint for getting/patching/deleting disaggregations (https://github.com/unicef/etools/pull/919)
- [x] (Vinod) Endpoint for listing/adding an intervention Reporting Period (#928)
- [x] (Vinod) Test Endpoint for listing/adding an intervention Reporting Period (#928)
- [x] (Vinod) Endpoint for patching/getting/deleting and intervention Reporting Period (#928)
- [x] (Vinod)Test Endpoint for patching/getting/deleting an intervention Reporting Period (#928)
- [x] (Csaba) Remove / migrate sector-locations to sectors only for the Intervention
- [x] Add API endpoint for PRP to GET Interventions
- [x] Add Permissions class for PRP endpoints to allow only get and only API users
- [x] Test API endpoint for PRP to get Interventions and test permissions class through it 

### Step 2 (this can be updated partially as we go along):
- [x] (Csaba) Refactor current InterventionDetail API response given the updated data models for result_links and lower ,  reporting periods , sector-locations, cluster-names,  location aggregation etc
- [x] Update the tests for the response of InterventionDetail to include ResultLinks and verify accuracy

- [x] Refactor the POST/PATCH api on the Intervention to remove the nested ResultLinks serializer. ResultLinks should not be updated through the api
- [x] Test the refactored POST/PATCH intervention API

- [x] + Endpoint for adding disaggregations by workspace
- [x] Test Endpoint for adding disaggregations by workspace

- [x] Address #883 
- [x] Address #934
...more to come

### CODE REVIEW

File: EquiTrack/.flake8 [hunk 1]
```diff
@@ -1,12 +1,6 @@
 [flake8]
 max-line-length = 120
-
-# Don't include inline comments in the exclude list. It might break the
-# exclude list on CircleCI due to a bad interaction between flake8 and
-# the 'future' library
-# https://gitlab.com/pycqa/flake8/issues/221
-# https://github.com/danicarrion/pyrestcli/issues/20
-
+ignore = F405,E121,E123,E126,E226,E24,E704,W503,W504
```
Comment (domdinicola) [inline EquiTrack/.flake8]: I think we have to remove this

File: EquiTrack/EquiTrack/admin.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+from snapshot.models import Activity
+from snapshot.utils import create_dict_with_relations, create_snapshot
+
+
+class SnapshotModelAdmin(admin.ModelAdmin):
+    def save_model(self, request, obj, form, change):
+        pre_save = None
+        if obj is not None:
+            if hasattr(obj, "pk") and obj.pk is not None:
+                pre_save = obj.__class__.objects.get(pk=obj.pk)
```
Comment (dpoirier) [inline EquiTrack/EquiTrack/admin.py]: ``obj.__class__`` could be changed to ``self.model`` which might read better.

File: EquiTrack/.flake8 (hunk 1 above)
Comment (domdinicola) [reply EquiTrack/.flake8]: @robertavram

File: EquiTrack/EquiTrack/admin.py (hunk 2 above)
Comment (domdinicola) [reply EquiTrack/EquiTrack/admin.py]: @robertavram

File: EquiTrack/EquiTrack/tests/test_all_models.py [hunk 3]
```diff
@@ -24,9 +23,9 @@
     'post_office',
 
     # These are the eTools packages that aren't yet using @python_2_unicode_compatible and therefore aren't yet
-    # Python 3-compatible. As they're fixed one by one, they'll be removed from this list.
-    'partners',
-    )
+    # Python 3-compatible. As they're fixed one by one, they'll be removed from this list
+    'users',
```
Comment (domdinicola) [inline EquiTrack/EquiTrack/tests/test_all_models.py]: isn't users compatible? @robertavram

File: EquiTrack/partners/views/interventions_v2.py [hunk 4]
```diff
@@ -351,17 +582,23 @@ class InterventionListMapView(ListCreateAPIView):
     permission_classes = (IsAdminUser,)
 
     def get_queryset(self):
-        q = Intervention.objects.detail_qs()\
-            .filter(sector_locations__isnull=False).exclude(sector_locations__locations=None)\
-            .prefetch_related('sector_locations__locations')
+        q = Intervention.objects.detail_qs()
+        # TODO: remember to add back the location filter after the PRP integration related structural changes are final
```
Comment (ewheeler) [inline EquiTrack/partners/views/interventions_v2.py]: 🤞

File: EquiTrack/partners/serializers/prp_v1.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    start_date = serializers.DateField(source='start')
+    end_date = serializers.DateField(source='end')
+    cso_budget = serializers.DecimalField(source='total_partner_contribution', read_only=True,
+                                          max_digits=20, decimal_places=2)
+    cso_budget_currency = serializers.CharField(source='default_budget_currency', read_only=True)
+    unicef_budget = serializers.DecimalField(source='total_unicef_budget', read_only=True,
+                                             max_digits=20, decimal_places=2)
+    unicef_budget_currency = serializers.CharField(source='default_budget_currency', read_only=True)
+    # todo: is this the right field?
+    funds_received = serializers.SerializerMethodField(read_only=True)
```
Comment (ewheeler) [inline EquiTrack/partners/serializers/prp_v1.py]: is this the right field?

File: EquiTrack/reports/models.py [hunk 6]
```diff
@@ -295,10 +372,16 @@ class Meta:
         ordering = ('-created',)
 
     def save(self, **kwargs):
+        # TODO: FIX THIS!
```
Comment (ewheeler) [inline EquiTrack/reports/models.py]: 🚧

File: EquiTrack/reports/serializers/v2.py [hunk 7]
```diff
@@ -28,22 +37,141 @@ class Meta:
...
+            DisaggregationValue.objects.create(disaggregation=disaggregation, **value_data)
+        return disaggregation
+
+    def update(self, instance, validated_data):
+        if instance.applied_indicators.count():
+            raise ValidationError(
+                'You cannot update a Disaggregation that is already associated with an Indicator.'
+            )
+        try:
+            values_data = validated_data.pop('disaggregation_values')
```
Comment (ewheeler) [inline EquiTrack/reports/serializers/v2.py]: flow control here and below seems too complex

File: EquiTrack/reports/serializers/v2.py [hunk 8]
```diff
@@ -28,22 +37,141 @@ class Meta:
...
+        section = attrs.get('section', None)
+        if section is None:
+            raise ValidationError(_('Section is required'))
+        if section.id not in [s.id for s in lower_result.result_link.intervention.sections.all()]:
+            raise ValidationError(_('This indicator can only have a section that was '
+                                    'previously saved on the intervention'))
+
+        if self.partial:
+            if not isinstance(blueprint_data, IndicatorBlueprint):
+                raise ValidationError(
```
Comment (ewheeler) [inline EquiTrack/reports/serializers/v2.py]: missing _ for i18n

File: EquiTrack/reports/serializers/v2.py [hunk 9]
```diff
@@ -28,22 +37,141 @@ class Meta:
...
+            indicator_blueprint = IndicatorBlueprintCUSerializer(data=blueprint_data)
+            indicator_blueprint.is_valid(raise_exception=True)
+            indicator_blueprint.save()
 
-    class Meta:
-        model = IndicatorBlueprint
-        fields = '__all__'
+            attrs["indicator"] = indicator_blueprint.instance
+            if lower_result.applied_indicators.filter(indicator__id=attrs['indicator'].id).exists():
+                raise ValidationError('This indicator is already being monitored for this Result')
```
Comment (ewheeler) [inline EquiTrack/reports/serializers/v2.py]: missing _ for i18n

**Pull request merged**